cedar-policy-validator/src/lib.rs

```diff
@@ -206,6 +286,192 @@ impl Validator {
         typecheck.typecheck_policy(t, &mut errors, &mut warnings);
         (errors.into_iter(), warnings.into_iter())
     }
+
+    #[cfg(feature = "level-validate")]
+    /// Check that `t` respects `max_allowed_level`
+    /// This assumes that (strict) typechecking has passed
+    fn check_entity_deref_level<'a>(
+        &'a self,
+        t: &'a Template,
+        max_allowed_level: &EntityDerefLevel,
+        policy_id: &PolicyID,
+    ) -> Vec<ValidationError> {
+        let typechecker = Typechecker::new(&self.schema, ValidationMode::Strict, t.id().clone());
+        let type_annotated_asts = typechecker.typecheck_by_request_env(t);
+        let mut errs = vec![];
+        for (_, policy_check) in type_annotated_asts {
+            match policy_check {
+                PolicyCheck::Success(e) => {
+                    let res =
+                        self.check_entity_deref_level_helper(&e, max_allowed_level, policy_id);
+                    match res.1 {
+                        Some(e) => errs.push(ValidationError::EntityDerefLevelViolation(e)),
+                        None => (),
+                    }
+                }
+                PolicyCheck::Fail(e) => println!("Typechecking failed: {:?}", e),
+                PolicyCheck::Irrelevant(_) => (), //Don't report level violations if typechecking already failed
+            }
+        }
+        errs
+    }
+
+    #[cfg(feature = "level-validate")]
+    fn min(
+        v: Vec<(EntityDerefLevel, Option<EntityDerefLevelViolation>)>,
+    ) -> (EntityDerefLevel, Option<EntityDerefLevelViolation>) {
+        match v.first() {
+            Some(e) => {
+                let mut p = e.clone();
+                for e in v {
+                    if e.0 < p.0 {
+                        p = e.clone();
+                    }
+                }
+                p
+            }
+            None => (EntityDerefLevel { level: None }, None),
+        }
+    }
+
+    #[cfg(feature = "level-validate")]
+    ///Walk the type-annotated AST and compute the used level and possible violation
```

Comment:
Can you add a comment here to explain how I should interpret the returned `EntityDerefLevel`.